Intake Sensors + Controls

Yumnah2 opened the pull request.

Yumnah2 committed:
Create Sensor Code

added sensor variables and some methods

Yumnah2 committed:
Merge branch 'robototes:main' into main

jamesdooley4 commented on src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java:
```diff
@@ -33,6 +34,11 @@ public class IntakeSubsystem extends SubsystemBase {
 	private final CANSparkFlex feederMotor;
 
+	// Sensors
+	private final DigitalInput intakeSensor;
```
The robot currently doesn't have an intake sensor, so should remove this & the getter method below

jamesdooley4 replied:
Rest of the code looks fine.  Nice work!

eddiert1239 replied:
Remove only the intake sensor and it's getter method, or the rest of them as well?

jamesdooley4 replied:
Just the intake sensor. The other sensors are good & should stay

SMKettle committed:
Added IntakeLowerMotor and Controls

based

TAKBS2412 commented on src/main/java/frc/team2412/robot/Hardware.java:
```diff
@@ -20,7 +20,13 @@ public class Hardware {
 	public static final int INTAKE_MOTOR_RIGHT = 43;
 
-	public static final int INDEX_MOTOR = 44;
+	public static final int INDEX_MOTOR_LOWER = 44;
+	public static final int INDEX_MOTOR_UPPER = 44;
```
These should be two different IDs.

eddiert1239 replied:
That wasn't very smart of us!

eddiert1239 commented:
Controls right now should be:
x button intakes in until the feeder sensor returns true, then all motors stop
b button stops all motors (intake index feeder)
y button intakes out while the button is held

jamesdooley4 commented on src/main/java/frc/team2412/robot/commands/intake/FeederOutCommand.java:
```diff
@@ -15,4 +15,9 @@ public FeederOutCommand(IntakeSubsystem intakeSubsystem) {
 	public void initialize() {
 		intakeSubsystem.feederOut();
 	}
+
+	@Override
+	public boolean isFinished() {
+		return false;
+	}
```
The feeder out, intake out & intake stop commands should all continue to return 'true' from isFinished (the defafult behavior, so can delete the override `isFinished` methods)

SMKettle committed:
Addressed changes

based

SMKettle committed:
created diagnostic command/all in command

made feeder index and intake in only intake in not using any sensor logic move that to all in until sensor command

TAKBS2412 commented on src/main/java/frc/team2412/robot/commands/intake/FeederInCommand.java:
```diff
@@ -15,14 +15,4 @@ public FeederInCommand(IntakeSubsystem intakeSubsystem) {
-	@Override
-	public boolean isFinished() {
-		return intakeSubsystem.feederSensor();
-	}
```
This part should be kept (same for `IndexInCommand` and `IntakeInCommand`) - what James and I meant was deleting `isFinished()` from the commands where it just returns `false`, like `FeederOutCommand`. Sorry for any confusion earlier.

jamesdooley4 replied:
Looking at [Command::isFinished](https://github.com/wpilibsuite/allwpilib/blob/da3abade83ada38d387d7ab0633f0914a44ff9d1/wpilibNewCommands/src/main/java/edu/wpi/first/wpilibj2/command/Command.java#L62), it returns `false` by default. I thought it returned `true`. So, the `isFinished()` override needs to be kept, and just updated to return `true` so the diagnostic command works correctly. Sorry about the confusion.

TAKBS2412 replied:
I'm fine with making commands like `IntakeStopCommand` and `IntakeOutCommand` return `true`, but for the "in" commands (`IndexInCommand`, `IntakeInCommand`, and `FeederInCommand`) I think we should keep `isFinished()` as is (checking the sensors) since the intent was to make the command run until a note is detected, which is additional logic I don't think we should compromise for a diagnostic command.

To keep the diagnostic command from being stuck waiting for a note, I'd suggest adding [`withTimeout()`](https://github.wpilib.org/allwpilib/docs/release/java/edu/wpi/first/wpilibj2/command/Command.html#withTimeout(double)) to make the in commands only run for a given period of time. (I could also see some value in keeping in commands that use the sensors, just to verify that the sensor hardware is working correctly).

Sorry for all the back and forth :(

jamesdooley4 replied:
Seems reasonable.

kirbt commented on src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java:
```diff
@@ -125,4 +137,13 @@ public void feederOut() {
 	}
+
+	// sensor methods
+	public boolean indexSensor() {
```
It would be nice to log these sensor getter methods for testing/debugging purposes

kirbt replied:
also the sensor method names are a bit iffy

SMKettle committed:
addressed changes

logging for sensors + changed isFinished to return true

TAKBS2412 commented on src/main/java/frc/team2412/robot/subsystems/IntakeSubsystem.java:
```diff
@@ -53,16 +59,32 @@ public class IntakeSubsystem extends SubsystemBase {
 
+	private final GenericEntry getIndexSensorEntry =
+			Shuffleboard.getTab("Intake")
+					.add("Index Sensor - ", getIndexSensor())
```
The `getIndexSensor()` call causes crashes (also in the `getFeederSensorEntry` below) because the index sensor hasn't been initialized yet. See #29 for a similar error that launcher subsystem ran into, and how to fix it (note you will have to update the entry's value in `periodic()`).

Alternatively, you can initialize the entries after the `DigitalInput`, and use something like [this](https://github.com/robototes/ChargedUp2023/blob/main/src/main/java/frc/team2412/robot/subsystems/VisionSubsystem.java#L165) to repeatedly call `getIndexSensor()` without using `periodic()`.

TAKBS2412 commented on src/main/java/frc/team2412/robot/Controls.java:
```diff
@@ -61,6 +83,28 @@ private void bindDrivebaseControls() {
+				Commands.race(
+						new IntakeInCommand(s.intakeSubsystem),
+						new IndexInCommand(s.intakeSubsystem),
+						new FeederInCommand(s.intakeSubsystem)));
```
Note that these and the parallel command groups below will throw an exception:

```
Error at frc.team2412.robot.Controls.bindIntakeControls(Controls.java:92): Unhandled exception: java.lang.IllegalArgumentException: Multiple commands in a parallel composition cannot require the same subsystems
        at edu.wpi.first.wpilibj2.command.ParallelRaceGroup.addCommands(ParallelRaceGroup.java:53)
        at edu.wpi.first.wpilibj2.command.ParallelRaceGroup.<init>(ParallelRaceGroup.java:35)
        at edu.wpi.first.wpilibj2.command.Commands.race(Commands.java:228)
        at frc.team2412.robot.Controls.bindIntakeControls(Controls.java:92)
        at frc.team2412.robot.Controls.<init>(Controls.java:69)
        at frc.team2412.robot.Robot.robotInit(Robot.java:67)
        at edu.wpi.first.wpilibj.TimedRobot.startCompetition(TimedRobot.java:107)
        at edu.wpi.first.wpilibj.RobotBase.runRobot(RobotBase.java:366)
        at edu.wpi.first.wpilibj.RobotBase.lambda$startRobot$0(RobotBase.java:438)
        at java.base/java.lang.Thread.run(Thread.java:833)
```

TAKBS2412 commented on src/main/java/frc/team2412/robot/Controls.java:
```diff
@@ -61,6 +83,28 @@ private void bindDrivebaseControls() {
+	private void bindIntakeControls() {
+		// later on set default command to check sensing notes
+		CommandScheduler.getInstance()
+				.setDefaultCommand(s.intakeSubsystem, new IntakeStopCommand(s.intakeSubsystem));
```
Because `IntakeOutCommand` ends immediately, the intake motors will only run briefly before `IntakeStopCommand` makes them all stop. I think not having a default command is ok for now especially since this one is just a placeholder.

eddiert1239 committed:
Added allin commands, hopefully fixed shuffleboard

all in commands to avoid add requirements issue with controls
removed the getFeederSensorEntry and replaced it with adding boolean in the constructor

eddiert1239 committed:
updated controls to what we talked about with driveteam

eddiert1239 commented:
### Controls now:
Drive controller:
x - in
y - out
b - stop
Codrive controller:
dpad left - in
dpad up - out
dpad right - stop

bobtseattle commented on src/main/java/frc/team2412/robot/commands/intake/AllOutCommand.java:
```diff
@@ -0,0 +1,25 @@
+import edu.wpi.first.wpilibj2.command.Command;
+import frc.team2412.robot.subsystems.IntakeSubsystem;
+
+public class AllOutCommand extends Command {
```
Out seems ambiguous.  It might spit the note out the bottom or out the launcher.

eddiert1239 committed:
Switched out with reverse

bobtseattle reviewed (approved):
LGTM
